File: python/2020/Day2.py
import sys
from common import DataAnalyzer

def passwd(data):
    for str in data:
        words = str.split()
        bounds = words[0].split('-')
        yield [int(bounds[0]), int(bounds[1]), words[1][0], words[2]]


def bound_check(data):
    valid = 0
    instructions = passwd(data)

    for i in instructions:
        match = i[3].count(i[2])
        if (match >= i[0] and match <= i[1]):
            valid += 1

    return valid


def index_check(data):
    valid = 0
    instructions = passwd(data)

    for i in instructions:
        if (i[3][i[0]-1] == i[2]) != (i[3][i[1]-1] == i[2]):
            valid += 1

    return valid


def second():
    data = [
        "1-3 a: abcde",
        "1-3 b: cdefg",
        "2-9 c: ccccccccc"
    ]
    print("(2.2.0) test valid passwords is {:d}".format(index_check(data)))

    data = DataAnalyzer.load("2020/day2.txt")
    print("(2.2) test valid passwords is {:d}".format(index_check(data)))


def first():
    data = [
        "1-3 a: abcde",
        "1-3 b: cdefg",
        "2-9 c: ccccccccc"
    ]

    print("(2.1.0) test valid passwords is {:d}".format(bound_check(data)))

    data = DataAnalyzer.load("2020/day2.txt")
    print("(2.1) valid passwords is {:}".format(bound_check(data)))


def solve(puzzle):
    if puzzle == '1':
        first()
    elif puzzle == '2':
        second()
    else:
        first()
        second()


if __name__ == '__main__':
    solve(sys.argv[1])

File: js/2020/Day2.js
const {DataAnalyzer} = require('../common/DataAnalyzer.js')
let data = DataAnalyzer.strs("2020/day2.txt")

function calc(interpreter=range) {
    let valid = 0;

    for (let i = 0; i <= data.length - 1; i++) {
        let entry = data[i];
        let tokens = entry.split(' ');
        let policy = tokens[0].split('-');

        valid += interpreter(policy[0], policy[1], tokens[1][0], tokens[2])
    }

    return valid;
}

function range(min, max, chr, passwd) {
    for (var n=count=0; n<passwd.length; count+=+(chr===passwd[n++]));
    if (count >= min && count <= max) {
        return 1;
    }

    return 0;
}

function pos(one, two, chr, passwd) {
    if (one > 0 && two > 0 && (passwd[one-1] === chr) !== (passwd[two-1] === chr)) {
        return 1;
    }

    return 0;
}

function one() {
    answer = calc()
    console.log("(2020.1): " + answer)
}

function two() {
    answer = calc(pos)
    console.log("(2020.2): " + answer)
}

switch(process.argv.slice(2)[0]) {
    case '1':
        one();
        break;

    case '2':
        two();
        break;
    
    default:
        one();
        two();
        break;
}